### PR TITLE
dird: Fix problem with reoccuring files in always incremental

### PR DESCRIPTION
**Pull request by arogge**
**Backport of PR #1395 to bareos-22**

#### Please check

- [x] Short description and the purpose of this PR is present _above this paragraph_
- [x] Your name is present in the AUTHORS file (optional)

If you have any questions or problems, please give a comment in the PR.

### Helpful documentation and best practices

- [Git Workflow](https://docs.bareos.org/DeveloperGuide/gitworkflow.html)
- [Automatic Sourcecode Formatting](https://docs.bareos.org/DeveloperGuide/generaldevel.html#automatic-sourcecode-formatting)
- [Check your commit messages](https://docs.bareos.org/DeveloperGuide/gitworkflow.html#commits)

### Checklist for the _reviewer_ of the PR (will be processed by the Bareos team)
Make sure you check/merge the PR using `devtools/pr-tool` to have some simple automated checks run and a proper changelog record added.

##### General
- [x] Is the PR title usable as CHANGELOG entry?
- [x] Purpose of the PR is understood
- [x] Commit descriptions are understandable and well formatted
- [x] Check backport line

##### Source code quality
- [x] Source code changes are understandable
- [x] Variable and function names are meaningful
- [x] Code comments are correct (logically and spelling)
- [x] Required documentation changes are present and part of the PR

##### Tests
- [x] Decision taken that a test is required (if not, then remove this paragraph)
- [x] The choice of the type of test (unit test or systemtest) is reasonable
- [x] Testname matches exactly what is being tested
- [x] On a fail, output of the test leads quickly to the origin of the fault

### CODE REVIEW

File: systemtests/tests/virtualfull-deletedfiles/testrunner-05-consolidate [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+quit
+END_OF_DATA
+
+run_bconsole
+
+# we expect 4 files: the added file, the changed directory and from each of the
+# two jobs a deleted file
+# however, with the catalog-only fix the deleted files will not be visible, so
+# we can only check for 2 files here.
+if ! grep -q 'Consolidating JobIds .* containing 2 files' "$tmp/$myname.log"; then
```
Comment (pstorz) [inline systemtests/tests/virtualfull-deletedfiles/testrunner-05-consolidate:38]: `expect_grep` again :)

File: systemtests/tests/virtualfull-deletedfiles/testrunner-05-consolidate [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+echo "$0" >"${tmp}/data/$myname"
+
+cat <<END_OF_DATA >"$tmp/bconcmds"
+@$out /dev/null
+messages
+@$out $tmp/$myname.log
+run job=backup-bareos-fd level=VirtualFull jobid=3,4 yes
+wait
+messages
+quit
```
Comment (pstorz) [inline systemtests/tests/virtualfull-deletedfiles/testrunner-05-consolidate:29]:
```suggestion
sqlquery                                                                           
select count(*) from file where jobid=5;
quit                                      
```

File: systemtests/tests/virtualfull-deletedfiles/testrunner-05-consolidate [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+if ! grep -q 'Replicating deleted files from jobids' "$tmp/$myname.log"; then
+  echo "Missing deleted file replication"
+  exit 2
+fi
+
+if ! grep -q 'SD Files Written:.*2' "$tmp/$myname.log"; then
+  echo "Number of files mismatch!"
+  exit 2
+fi
+
```
Comment (pstorz) [inline systemtests/tests/virtualfull-deletedfiles/testrunner-05-consolidate:51]:
```suggestion
expect_grep '^| *4 |' "$tmp/$myname.log" "Number of files including fileindex 0 mismatch"
```

Comment (pstorz) [reply systemtests/tests/virtualfull-deletedfiles/testrunner-05-consolidate:51]: This is to make sure that the two deleted files really where replicated.

**Review by pstorz — CHANGES_REQUESTED** [review]
See comments.

File: systemtests/tests/reload/CMakeLists.txt [hunk 4]
```diff
@@ -19,15 +19,3 @@
 
 get_filename_component(BASENAME ${CMAKE_CURRENT_BINARY_DIR} NAME)
 create_systemtest(${SYSTEMTEST_PREFIX} ${BASENAME})
-
```
[the remark this answers is not in the document]
Comment (sebsura) [reply systemtests/tests/reload/CMakeLists.txt:22]: yes. we fixed that issue.  In fact only add-second-director was broken anyways.

File: core/src/dird/vbackup.cc [hunk 5]
```diff
@@ -452,6 +439,27 @@ void NativeVbackupCleanup(JobControlRecord* jcr, int TermCode, int JobLevel)
...
+        jcr->dir_impl->vf_jobids, jcr->dir_impl->vf_jobids);
+    std::string query
+        = "INSERT INTO File (FileIndex, JobId, PathId, LStat, MD5, Name) "s
+          + "SELECT FileIndex, "s + std::to_string(jcr->JobId) + " AS JobId, "s
+          + "PathId, LStat, MD5, Name FROM ("s + q1.c_str() + ") T "s
+          + "WHERE FileIndex = 0"s;
+    if (!jcr->db->SqlQuery(query.c_str())) {
+      Jmsg(jcr, M_WARNING, 0, "Error replicating deleted files: ERR=%s\n",
+           jcr->db->strerror());
+    }
```
[the remark this answers is not in the document]
Comment (arogge) [reply core/src/dird/vbackup.cc:460]: as discussed yesterday that's not really possible to do.

**Pull request merged**